publisher-sdk/src/main/java/com/criteo/publisher/util/jsonadapter/ArrayListJsonAdapter.kt

```diff
@@ -0,0 +1,108 @@
+/*
+ *    Copyright 2020 Criteo
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+package com.criteo.publisher.util.jsonadapter
+
+import com.squareup.moshi.JsonAdapter
+import com.squareup.moshi.JsonReader
+import com.squareup.moshi.JsonWriter
+import com.squareup.moshi.Moshi
+import com.squareup.moshi.Types
+import java.io.IOException
+import java.lang.reflect.Type
+import java.util.Arrays
+
+internal abstract class ArrayListJsonAdapter<C : MutableCollection<T>?, T> private constructor(
+    private val elementAdapter: JsonAdapter<T>
+) : JsonAdapter<C>() {
+  abstract fun newCollection(): C
+
+  @Throws(IOException::class)
+  override fun fromJson(reader: JsonReader): C {
+    val result = newCollection()
+    reader.beginArray()
+    while (reader.hasNext()) {
+      result?.add(elementAdapter.fromJson(reader)!!)
+    }
+    reader.endArray()
+    return result
+  }
+
+  @Throws(IOException::class)
+  override fun toJson(writer: JsonWriter, value: C?) {
+    writer.beginArray()
+    for (element in value!!) {
+      elementAdapter.toJson(writer, element)
+    }
+    writer.endArray()
+  }
+
+  override fun toString(): String {
+    return "$elementAdapter.collection()"
+  }
+
+  companion object {
+    @JvmStatic
+    val factory = Factory { type, annotations, moshi ->
+      val rawType = Types.getRawType(type)
+      if (annotations.isNotEmpty()) return@Factory null
+      // not really the best way to compare but class is private
+      if (rawType.name == "java.util.Arrays${'$'}ArrayList") {
+        return@Factory newArraysArrayListAdapter<Any>(
+            type,
+            moshi
+        ).nullSafe()
+      }
+      if (rawType == java.util.ArrayList::class.java) {
+        return@Factory newArrayListAdapter<Any>(
+            type,
+            moshi
+        ).nullSafe()
+      }
+      null
+    }
+
+    private fun <T> newArraysArrayListAdapter(
+        type: Type,
+        moshi: Moshi
+    ): JsonAdapter<MutableCollection<T>> {
+      val elementType = Types.collectionElementType(type, Collection::class.java)
+      val elementAdapter: JsonAdapter<T> = moshi.adapter(elementType)
+
+      return object :
+          ArrayListJsonAdapter<MutableCollection<T>, T>(elementAdapter) {
+        override fun newCollection(): MutableCollection<T> {
+          return Arrays.asList()
```

Comment:
AFAIK, Arrays.asList is not mutable and this could generate exception at runtime

Reply:
Yep, you are right. Actually we need this class only for object to json conversion. At runtime when we get moshi adapter it is fetched for a runtime class. This can be fixed by converting serialization code to kotlin and using inline function with reified generic. But doing so requires a lot of changes in code and we will need to get moshi statically from inline function.

I will write some comments on when this class is used